Support for Multiple Keywords

Dayjo opened the pull request:
This pull request implements the ability to use an array of keywords as well as just a single string. 

This doesn't update any documentation, I want to make sure that the code is acceptable first!

Any suggestions / improvements I can make please let me know @pat310 

Fixes #53

Dayjo commented:
Any assistance on the code coverage would be appreciated, not 100% sure what I need to do to fix it.

pat310 commented:
@Dayjo Cool! I think we should move this whole thing to a new function in `utilites.js` and that will return the `comparisonItem` array.  `getResults` will call that function at line 116.  Then add some tests for that new function in `utilities.spec.js`, that should get the code coverage back up.  Let me know if you need help

pat310 commented on src/utilities.js:
```diff
@@ -107,13 +107,34 @@ export function getResults(searchType, obj) {
+  var comparisonItems = [];
+
+  // If we are requesting an array of keywords for comparison
+  if (typeof obj.keyword === 'object' && obj.keyword.length) {
```
Check if it's an array using `Array.isArray(obj.keyword)`

pat310 commented on src/utilities.js:
```diff
@@ -107,13 +107,34 @@ export function getResults(searchType, obj) {
+      obj2.keyword = keyword;
+
+      // Add the keyword to the array
+      comparisonItems.push(obj2);
```
Rather than having a side effect within this `.map`, use a `.reduce` on `obj.keyword` to create a new array populated with the original object and the keywords replaced.  For example:
```js
comparisonItem = obj.keyword.reduce((arr, keyword) => {
 const copiedObj = {...obj, keyword};

 arr.push(copiedObj);
 return arr;
}, [])
```

Dayjo replied:
Strangely I'm getting errors on the object spread syntax.

```js
ERROR in ./src/utilities.js
Module build failed: SyntaxError: /google-trends-api/node_modules/eslint-loader/index.js!/google-trends-api/src/utilities.js: Unexpected token (99:16)
   97 |     obj.keyword.reduce((arr, keyword) => {
   98 |       // Add the keyword to the array
>  99 |       arr.push({...obj,keyword});
      |                 ^
```

pat310 commented on src/utilities.js:
```diff
@@ -107,13 +107,34 @@ export function getResults(searchType, obj) {
+    obj.keyword.map((keyword) => {
+
+      // Clone the original obj to keep the dates etc
+      let obj2 = JSON.parse(JSON.stringify(obj));
```
I'm following [airbnb's javascript](https://github.com/airbnb/javascript#objects--rest-spread) guide.  Instead of doing `JSON.parse(JSON.stringify` to copy an object, use `{...obj}` (should be fine because  we only need a shallow copy)

pat310 commented on src/utilities.js:
```diff
@@ -107,13 +107,34 @@ export function getResults(searchType, obj) {
   };
 
+  // Create array of items to query
+  var comparisonItems = [];
```
use `let` and change the name to `comparisonItem` so that way we can use shorthand notation below:
`let comparisonItem;`

pat310 commented on src/utilities.js:
```diff
@@ -107,13 +107,34 @@ export function getResults(searchType, obj) {
     qs: {
       hl: obj.hl,
-      req: JSON.stringify({comparisonItem: [obj], cat: 0}),
+      req: JSON.stringify({comparisonItem: comparisonItems, cat: 0}),
```
Then here we can use shorthand notation:
`req: JSON.stringify({comparisonItem, cat: 0}),`

Dayjo commented:
Thanks for this is, great feedback thank you 👍

Dayjo commented:
@pat310 inline comments have disappeared cause I pushed my latest changes (failing). 

As you can see, the syntax checker is failing on the `{...obj` part unfortunately so I can't build/test.

Dayjo commented:
@pat310 I added `"transform-es2015-destructuring", "transform-object-rest-spread"` to the babel plugins to fix the syntax checking for object spread.

Also added two tests for the `formatKeywords` function  one to check for normal single keyword and one for checking an array of keywords. Do you have any suggestions for what else I can test on this?

Dayjo commented:
Eh, not sure why the coverage has dropped, pretty sure my new function is covered.

pat310 commented on src/utilities.js:
```diff
@@ -86,6 +86,26 @@ export function formatResolution(resolution = '') {
   return '';
 }
 
+export function formatKeywords(obj) {
```
Cool! :+1:

pat310 commented on src/utilities.js:
```diff
@@ -86,6 +86,26 @@ export function formatResolution(resolution = '') {
+  var comparisonItems = [obj];
+
+  // If we are requesting an array of keywords for comparison
+  if (Array.isArray(obj.keyword)) {
```
Having `comparisonItems` variable is kind of unnecessary.  Maybe instead start with the `if (Array.isarray(obj.keyword))` check and inside you can simply return the reduce.

pat310 commented on src/utilities.js:
```diff
@@ -86,6 +86,26 @@ export function formatResolution(resolution = '') {
+
+  // If we are requesting an array of keywords for comparison
+  if (Array.isArray(obj.keyword)) {
+    comparisonItems = []; // Reset to empty
```
Get rid of this line

pat310 commented on src/utilities.js:
```diff
@@ -86,6 +86,26 @@ export function formatResolution(resolution = '') {
+    comparisonItems = []; // Reset to empty
+
+    // Map the keywords to the items array
+    obj.keyword.reduce((arr, keyword) => {
```
```js
return obj.keyword.reduce((arr, keyword) => {
....
```

pat310 commented on src/utilities.js:
```diff
@@ -86,6 +86,26 @@ export function formatResolution(resolution = '') {
+      arr.push({ ...obj, keyword });
+
+      return arr;
+    }, comparisonItems);
```
Initiate reduce with empty array:
```js
...
return arr;
}, []);
```

pat310 commented on src/utilities.js:
```diff
@@ -107,13 +127,15 @@ export function getResults(searchType, obj) {
   };
 
+  // Create an array of keywords to query
+  let comparisonItem = formatKeywords(obj);
```
Now we don't need the `comparisonItem` variable here either, just plug `formatKeywords(obj)` right into the `JSON.stringify(`.

So get rid of this line

pat310 commented on src/utilities.js:
```diff
@@ -107,13 +127,15 @@ export function getResults(searchType, obj) {
     qs: {
       hl: obj.hl,
-      req: JSON.stringify({comparisonItem: [obj], cat: 0}),
+      req: JSON.stringify({comparisonItem, cat: 0}),
```
And then here:
`re: JSON.stringify({comparisonItem: formatKeywords(obj), cat: 0}),`

pat310 commented on src/utilities.js:
```diff
@@ -86,6 +86,26 @@ export function formatResolution(resolution = '') {
+
+  }
+
+  return comparisonItems;
```
Here then we can just return obj in an array:
`return [obj];`

pat310 commented on test/utilities.spec.js:
```diff
@@ -80,6 +81,18 @@ describe('utilities', () => {
 
+  describe('formatKeywords', () => {
+    it('should return single comparisonItem', () => {
+      let keywords = formatKeywords({ keyword: 'test' });
```
Add a blank line beneath variable declarations

pat310 commented:
Looks good @Dayjo.  Just a few last comments and I think it's good to go

pat310 commented:
Do you want to update the `README` as well?

turnerniles commented:
@Dayjo Are you nightcrawler in your profile picture?

Dayjo commented:
Tests seem to pass but I do get an error here;

![image](https://cloud.githubusercontent.com/assets/1255512/23792725/20699c20-0580-11e7-8c90-39da8d5bff3f.png)
